trieste/utils/misc.py

```diff
@@ -235,3 +235,21 @@ def __exit__(
     ) -> None:
         self.end = perf_counter()
         self.time = self.end - self.start
+
+
+def flatten_leading_dims(x: TensorType) -> Tuple[TensorType, Callable[[TensorType], TensorType]]:
+    """
+    Flattens the leading dimensions of `x`, and returns a function that can be used to restore them.
```

Comment:
Maybe an extra few words explain why this is useful?

```suggestion
    Flattens the leading dimensions of `x`, and returns a function that can be used to restore them
    (typically after first manipulating the flattened tensor).
```